risc0/zkvm/src/host/receipt.rs

```diff
@@ -206,14 +214,29 @@ impl SegmentReceipts {
         final_receipt.verify_with_context(ctx)?;
         let metadata = final_receipt.get_metadata()?;
         log::debug!("final: {metadata:#?}");
-        if prev_image_id != metadata.pre.digest() {
+        // If there's more than one segment, the last segment could be the fault checker
+        if prev_image_id != metadata.pre.digest()
+            && !(cfg!(feature = "enable-fault-proof")
+                && receipts.len() > 0
+                && metadata.pre.digest() == FAULT_CHECKER_ID.into())
```

Comment:
It took me a minute to parse this correctly. Breaking out the complicated inner logic would be helpful.

Also, if `receipts.len() == 0`, then `prev_image_id` will be equal to `metadata.pre.digest()`, so we don't need to check that condition I don't think.
```suggestion
        let final_receipt_is_fault_checker = cfg!(feature = "enable-fault-proof")
                && metadata.pre.digest() == FAULT_CHECKER_ID.into();
        if prev_image_id != metadata.pre.digest()
            && !final_receipt_is_fault_checker
```